test/ng/parseSpec.js

```diff
@@ -2840,6 +2840,46 @@ describe('parser', function() {
             expect(filterCalled).toBe(true);
           });
 
+          it('should not be invoked unless the input/arguments change within literals', function() {
+            var filterCalls = [];
+            $filterProvider.register('foo', valueFn(function(input) {
+              filterCalls.push(input);
+              return input;
+            }));
+
+            scope.$watch('[(a | foo:b:1), undefined]');
+            scope.a = 0;
+            scope.$digest();
+            expect(filterCalls).toEqual([0]);
+
+            scope.$digest();
+            expect(filterCalls).toEqual([0]);
+
+            scope.a++;
+            scope.$digest();
+            expect(filterCalls).toEqual([0, 1]);
+          });
+
+          it('should not be invoked unless the input/arguments change within literals (one-time)', function() {
+            var filterCalls = [];
+            $filterProvider.register('foo', valueFn(function(input) {
+              filterCalls.push(input);
+              return input;
+            }));
+
+            scope.$watch('::[(a | foo:b:1), undefined]');
+            scope.a = 0;
+            scope.$digest();
+            expect(filterCalls).toEqual([0]);
+
+            scope.$digest();
+            expect(filterCalls).toEqual([0]);
+
+            scope.a++;
+            scope.$digest();
+            expect(filterCalls).toEqual([0, 1]);
```

Comment:
This seems unexpected. If this is a one-time binding, shouldn't we be ignoring the filter once the binding has a value?

Reply:
Or does the `undefined` in the second element of the array prevent the one time binding from stabilizing?

Reply:
Exactly! For `::[ ]` and `::{ }`, all values must be defined before unwatching the whole expression.